pkg/lib/cockpit/dialog.tsx

```diff
@@ -0,0 +1,1543 @@
+       to wait for those timeouts to be over from here.)
+     */
+
+    async validate(): Promise<boolean> {
+        this.#cancel_all_validation_timeouts();
+        this.#trigger_validation();
+        await this.#wait_for_validation_promises();
+        return !this.#validation_failed;
+    }
+
+    #cancel_all_validation_timeouts() {
+        for (const p in this.#validation_state) {
+            const state = this.#validation_state[p];
+            if (state.timeout_id) {
+                debug("timeout bulk cancel", p);
+                window.clearTimeout(state.timeout_id);
+                delete this.#validation_state[p];
+            }
+        }
+    }
+
+    async #wait_for_validation_promises(): Promise<void> {
+        for (const path in this.#validation_state) {
+            const state = this.#validation_state[path];
+            if (state.promise) {
+                debug("waiting for promise", path);
+                await state.promise;
+                debug("waiting for promise done", path);
+            }
+        }
+    }
+
+    async run_action(func: (vals: V) => Promise<void>) {
+        this.error = null;
+        this.#action_running = true;
+        this.#update();
+        if (!await this.validate()) {
+            this.#action_running = false;
+            this.#update();
+            return;
+        }
+
+        try {
+            await func(this.values);
+        } catch (ex) {
+            console.error(String(ex));
+            this.error = ex;
+        }
+
+        this.#action_running = false;
+        this.#update();
+    }
+
+    top_value(update_func?: ((val: V) => void) | undefined): DialogValue<V> {
+        return new DialogValue<V>(
+            this as DialogState<unknown>,
+            () => this.values,
+            (val) => {
+                debug("set", val);
+                if (this.#action_running) {
+                    debug("set denied");
+                    return;
+                }
+                this.values = val;
+                this.#update();
+                if (this.#online_validation)
+                    this.#trigger_validation();
+                if (update_func)
+                    update_func(val);
+            },
+            "");
+    }
+
+    value<K extends keyof V>(tag: K, update_func?: ((val: V[K]) => void) | undefined): DialogValue<V[K]> {
+        return this.top_value().sub(tag, update_func);
+    }
+}
+
+export class DialogError {
+    title: string;
+    details: React.ReactNode;
+
+    constructor(title: string, details: React.ReactNode) {
+        this.title = title;
+        this.details = details;
+    }
+
+    toString() {
+        return this.title + ": " + String(this.details);
+    }
+
+    static fromError(title: string, err: unknown) {
+        if (err && typeof err == "object" && "message" in err && typeof err.message == "string")
+            return new DialogError(title, err.message);
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22700-2dd5f8c0-20251219-103022-fedora-43-devel-other/Coverage/lcov/github-pr.diff.gcov.html)